exploration/bidi-usability.md

```diff
@@ -348,47 +407,51 @@ the results or debug what is wrong with their messages.
+This would permit users to get the effects described by the above design,
+as long as they use isolates/marks in a "responsible" way.
 
-### Loose isolation
+(Omitting other changes found in #673)
 
-Apply bidi isolates in a slightly different way.
-The main differences to the proposed solution are:
-1. The open/close isolate characters are not syntactically required to be paired.
-   This avoids introducing parse errors for missing or required invisible characters,
-   which would lead to bad user experiences.
-2. Rather than patching the `name` rule with an optional trailing LRM/RLM/ALM,
-   allow for its proper isolation.
+```abnf
+; strongly directional marks and bidi isolates
+; ALM / LRM / RLM / LRI / RLI / FSI / PDI
+bidi = %x061C / %x200E / %x200F / %x2066-2069
 
-Quoted patterns, quoted literals, and names may be isolated by LRI/RLI/FSI...PDI.
-For names and quoted literals, the isolate characters are outside the body of the token,
-but for quoted patterns, the isolates are in the middle of the `{{` and `}}` characters.
-This avoids adding a lookahead requirement for detecting a `complex-message` start,
-and differentiates a `quoted-pattern` from a `quoted` `key` in a `variant`.
+; optional whitespace
+owsp = *( s / bidi )
 
-Expressions and markup may be isolated by LRI...PDI immediately within the `{` and `}`.
+; required whitespace
+wsp = [ owsp ] 1*s [ owsp ]
 
-An LRI is allowed immediately after a newline outside patterns and within expressions.
-This is intended to allow left-to-right representation for "code"
-even if it contains a newline followed by content
-that could otherwise prompt the paragraph direction to be detected as right-to-left.
+; whitespace characters
+s = ( SP / HTAB / CR / LF / %x3000 )
+```
 
-```abnf
-name           = [open-isolate] name-start *name-char [close-isolate]
-quoted         = [open-isolate] "|" *(quoted-char / quoted-escape) "|" [close-isolate]
-quoted-pattern = "{" [open-isolate] "{" pattern "}" [close-isolate] "}"
+**Pros**
+- Avoids problems with syntax errors that users and tools might find difficult to debug.
+- Effective if used carefully.
+- Addresses need to comply with UAX#31
 
-literal-expression    = "{" [LRI] [s] literal [s annotation] *(s attribute) [s] [close-isolate] "}"
-variable-expression   = "{" [LRI] [s] variable [s annotation] *(s attribute) [s] [close-isolate] "}"
-annotation-expression = "{" [LRI] [s] annotation *(s attribute) [s] [close-isolate] "}"
+**Cons**
+- Syntax does not prevent poor display outcomes, including enabling some Trojan Source cases (UAX#55);
+  note that tooling or linting can help ameliorate these issues.
 
-markup = "{" [LRI] [s] "#" identifier *(s option) *(s attribute) [s] ["/"] [close-isolate] "}"
-       / "{" [LRI] [s] "/" identifier *(s option) *(s attribute) [s] [close-isolate] "}"
+### Strict isolation all the time
+
+Apply bidi isolates in a strict way.
+The main differences to the proposed solution is:
```

Comment:
I only see one difference?

Reply:
I should do the TODO, which makes it clearer. I'll do one here in the comment for clarity and then go back and fix the PR.

The current design has `expression` thusly:

```abnf
expression = "{" [LRI] (literal-expression / variable-expression / annotation-expression) [close-isolate] "}"
```

This alternative would turn that into:

```abnf
expression = "{" (literal-expression / variable-expression / annotation-expression) "}"
           / "{" LRI (literal-expression / variable-expression / annotation-expression) close-isolate "}"
```

In this formulation, you cannot have unpaired opening (or closing) isolates without a syntax error, nor can you have multiples of open or close.

Rinse and repeat for markup, option, attribute, and literals.

Make sense?